File: strategies/3-greedy/ActivitySelection.c
#include <stdio.h>
#include <stdlib.h>

/**
 * Greedy Strategy: Activity Selection Problem
 * Core Idea: Always choose the activity that finishes earliest (and doesn't conflict)
 * Time Complexity: O(n log n) due to sorting by finish time
 * Space Complexity: O(1) excluding input
 */

/**
 * Activity structure
 */
typedef struct {
    int start;
    int finish;
    int index; // Original index for tracking
    char name[10]; // Activity name for display
} Activity;

/**
 * Compare function for sorting activities by finish time
 */
int compareByFinishTime(const void* a, const void* b) {
    Activity* actA = (Activity*)a;
    Activity* actB = (Activity*)b;
    return actA->finish - actB->finish;
}

/**
 * Solve activity selection using greedy approach
 * @param activities Array of activities
 * @param n Number of activities
 * @param selected Array to store selection status
 * @return Number of selected activities
 */
int activitySelection(Activity activities[], int n, int selected[]) {
    if (n == 0) return 0;
    
    // Step 1: Sort activities by finish time (greedy choice preparation)
    qsort(activities, n, sizeof(Activity), compareByFinishTime);
    
    // Initialize selection array
    for (int i = 0; i < n; i++) {
        selected[activities[i].index] = 0;
    }
    
    // Step 2: Always select the first activity (earliest finish time)
    selected[activities[0].index] = 1;
    int count = 1;
    int lastSelectedFinishTime = activities[0].finish;
    
    // Step 3: Greedily select non-conflicting activities
    for (int i = 1; i < n; i++) {
        // Greedy choice: if activity starts after last selected activity finishes
        if (activities[i].start >= lastSelectedFinishTime) {
            selected[activities[i].index] = 1;
            lastSelectedFinishTime = activities[i].finish;
            count++;
        }
    }
    
    return count;
}

/**
 * Check if two activities conflict
 */
int hasConflict(Activity a, Activity b) {
    return !(a.finish <= b.start || b.finish <= a.start);
}

/**
 * Print activities information
 */
void printActivities(Activity originalActivities[], int n) {
    printf("Activities (Name, Start, Finish):\n");
    for (int i = 0; i < n; i++) {
        printf("  %s: [%d, %d]\n", 
               originalActivities[i].name, 
               originalActivities[i].start, 
               originalActivities[i].finish);
    }
}

/**
 * Print selection result
 */
void printSelection(Activity originalActivities[], int n, int selected[], int count) {
    printf("Selected Activities (%d total):\n", count);
    
    int selectionOrder = 1;
    for (int i = 0; i < n; i++) {
        if (selected[i]) {
            printf("  %d. %s: [%d, %d]\n", 
                   selectionOrder++, originalActivities[i].name,
                   originalActivities[i].start, originalActivities[i].finish);
        }
    }
}

/**
 * Demonstrate greedy algorithm step by step
 */
void demonstrateGreedySteps(Activity activities[], int n) {
    printf("Greedy Algorithm Steps:\n");
    
    // Create copy for sorting
    Activity* sortedActivities = (Activity*)malloc(n * sizeof(Activity));
    for (int i = 0; i < n; i++) {
        sortedActivities[i] = activities[i];
    }
    
    qsort(sortedActivities, n, sizeof(Activity), compareByFinishTime);
    
    printf("1. Sort by finish time:\n");
    for (int i = 0; i < n; i++) {
        printf("   %s: [%d, %d]\n", 
               sortedActivities[i].name, 
               sortedActivities[i].start, 
               sortedActivities[i].finish);
    }
    
    printf("2. Greedy selection process:\n");
    printf("   - Select %s (first activity, finishes at %d)\n", 
           sortedActivities[0].name, sortedActivities[0].finish);
    
    int lastFinish = sortedActivities[0].finish;
    int count = 1;
    
    for (int i = 1; i < n; i++) {
        if (sortedActivities[i].start >= lastFinish) {
            printf("   - Select %s (starts at %d >= %d, finishes at %d)\n", 
                   sortedActivities[i].name, sortedActivities[i].start, 
                   lastFinish, sortedActivities[i].finish);
            lastFinish = sortedActivities[i].finish;
            count++;
        } else {
            printf("   - Skip %s (starts at %d < %d, conflicts with previous)\n", 
                   sortedActivities[i].name, sortedActivities[i].start, lastFinish);
        }
    }
    
    printf("3. Total selected: %d activities\n", count);
    free(sortedActivities);
}

/**
 * Create timeline visualization
 */
void printTimeline(Activity originalActivities[], int n, int selected[]) {
    printf("Timeline Visualization:\n");
    
    // Find time range
    int minTime = originalActivities[0].start;
    int maxTime = originalActivities[0].finish;
    
    for (int i = 1; i < n; i++) {
        if (originalActivities[i].start < minTime) minTime = originalActivities[i].start;
        if (originalActivities[i].finish > maxTime) maxTime = originalActivities[i].finish;
    }
    
    // Print time scale
    printf("Time: ");
    for (int t = minTime; t <= maxTime; t++) {
        printf("%2d ", t);
    }
    printf("\n");
    
    // Print each activity
    for (int i = 0; i < n; i++) {
        printf("%-4s: ", originalActivities[i].name);
        
        for (int t = minTime; t <= maxTime; t++) {
            if (t >= originalActivities[i].start && t < originalActivities[i].finish) {
                printf(selected[i] ? " * " : " - ");
            } else {
                printf("   ");
            }
        }
        
        printf(" %s\n", selected[i] ? "(Selected)" : "(Skipped)");
    }
    
    printf("Legend: * = Selected activity, - = Unselected activity\n");
}

/**
 * Compare with alternative selection strategy
 */
void compareStrategies(Activity activities[], int n) {
    printf("Strategy Comparison:\n");
    
    // Greedy by finish time
    Activity* greedyActivities = (Activity*)malloc(n * sizeof(Activity));
    for (int i = 0; i < n; i++) {
        greedyActivities[i] = activities[i];
    }
    
    int* greedySelected = (int*)calloc(n, sizeof(int));
    int greedyCount = activitySelection(greedyActivities, n, greedySelected);
    
    printf("1. Greedy by finish time: %d activities\n", greedyCount);
    
    // Alternative: Greedy by duration (shortest first)
    printf("2. Alternative - Greedy by shortest duration:\n");
    
    // Find shortest duration activities
    int shortestCount = 0;
    int* shortestSelected = (int*)calloc(n, sizeof(int));
    
    // Simple greedy by shortest duration (not optimal, just for comparison)
    int used[n];
    for (int i = 0; i < n; i++) used[i] = 0;
    
    while (1) {
        int shortestIdx = -1;
        int shortestDuration = 1000;
        
        // Find shortest unused activity that doesn't conflict
        for (int i = 0; i < n; i++) {
            if (used[i]) continue;
            
            int duration = activities[i].finish - activities[i].start;
            if (duration < shortestDuration) {
                // Check conflicts with already selected
                int hasConflictWithSelected = 0;
                for (int j = 0; j < n; j++) {
                    if (shortestSelected[j] && hasConflict(activities[i], activities[j])) {
                        hasConflictWithSelected = 1;
                        break;
                    }
                }
                
                if (!hasConflictWithSelected) {
                    shortestIdx = i;
                    shortestDuration = duration;
                }
            }
        }
        
        if (shortestIdx == -1) break; // No more activities can be added
        
        shortestSelected[shortestIdx] = 1;
        used[shortestIdx] = 1;
        shortestCount++;
    }
    
    printf("   Selected %d activities (may not be optimal)\n", shortestCount);
    printf("   Greedy by finish time is provably optimal!\n");
    
    free(greedyActivities);
    free(greedySelected);
    free(shortestSelected);
}

int main() {
    printf("=== Activity Selection - Greedy Algorithm ===\n");
    
    // Test Case 1: Classic example
    printf("Test Case 1: Classic activity selection\n");
    Activity activities1[] = {
        {1, 4, 0, "A1"},
        {3, 5, 1, "A2"},
        {0, 6, 2, "A3"},
        {5, 7, 3, "A4"},
        {3, 9, 4, "A5"},
        {5, 9, 5, "A6"},
        {6, 10, 6, "A7"},
        {8, 11, 7, "A8"},
        {8, 12, 8, "A9"},
        {2, 14, 9, "A10"},
        {12, 16, 10, "A11"}
    };
    int n1 = 11;
    int selected1[11];
    
    printActivities(activities1, n1);
    printf("\n");
    
    int count1 = activitySelection(activities1, n1, selected1);
    printSelection(activities1, n1, selected1, count1);
    printf("\n");
    
    demonstrateGreedySteps(activities1, n1);
    printf("\n");
    
    printTimeline(activities1, n1, selected1);
    printf("\n");
    
    // Test Case 2: Simple non-overlapping case
    printf("Test Case 2: Simple non-overlapping activities\n");
    Activity activities2[] = {
        {1, 3, 0, "B1"},
        {4, 6, 1, "B2"},
        {7, 9, 2, "B3"}
    };
    int n2 = 3;
    int selected2[3];
    
    printActivities(activities2, n2);
    printf("\n");
    
    int count2 = activitySelection(activities2, n2, selected2);
    printSelection(activities2, n2, selected2, count2);
    printf("Note: All activities selected since none overlap\n\n");
    
    // Test Case 3: All activities overlap
    printf("Test Case 3: All activities overlap\n");
    Activity activities3[] = {
        {1, 5, 0, "C1"},
        {2, 6, 1, "C2"},
        {3, 7, 2, "C3"},
        {2, 8, 3, "C4"}
    };
    int n3 = 4;
    int selected3[4];
    
    printActivities(activities3, n3);
    printf("\n");
    
    int count3 = activitySelection(activities3, n3, selected3);
    printSelection(activities3, n3, selected3, count3);
    printf("Note: Only one activity selected since all overlap\n");
    printf("Greedy chooses the one that finishes earliest (C1)\n\n");
    
    // Test Case 4: Edge case - single activity
    printf("Test Case 4: Single activity\n");
    Activity activities4[] = {
        {2, 5, 0, "D1"}
    };
    int n4 = 1;
    int selected4[1];
    
    printActivities(activities4, n4);
    printf("\n");
    
    int count4 = activitySelection(activities4, n4, selected4);
    printSelection(activities4, n4, selected4, count4);
    printf("\n");
    
    // Test Case 5: Strategy comparison
    printf("Test Case 5: Comparing selection strategies\n");
    Activity activities5[] = {
        {0, 2, 0, "E1"},  // Duration: 2
        {1, 100, 1, "E2"}, // Duration: 99 (long, blocks many)
        {3, 4, 2, "E3"},  // Duration: 1
        {5, 6, 3, "E4"},  // Duration: 1
        {7, 8, 4, "E5"}   // Duration: 1
    };
    int n5 = 5;
    int selected5[5];
    
    printActivities(activities5, n5);
    printf("\n");
    
    int count5 = activitySelection(activities5, n5, selected5);
    printSelection(activities5, n5, selected5, count5);
    printf("\n");
    
    compareStrategies(activities5, n5);
    
    return 0;
}

File: strategies/3-greedy/ActivitySelection.java
import java.util.*;

/**
 * Greedy Strategy: Activity Selection Problem
 * Core Idea: Always choose the activity that finishes earliest and doesn't conflict
 * Time Complexity: O(n log n) due to sorting by finish time
 * Space Complexity: O(n) for storing activities and results
 */
public class ActivitySelection {
    
    /**
     * Activity class to represent activities with start and finish times
     */
    static class Activity {
        int id;
        int startTime;
        int finishTime;
        String name;
        
        Activity(int id, String name, int startTime, int finishTime) {
            this.id = id;
            this.name = name;
            this.startTime = startTime;
            this.finishTime = finishTime;
        }
        
        @Override
        public String toString() {
            return String.format("A%d:%s [%d-%d]", id, name, startTime, finishTime);
        }
    }
    
    /**
     * Select maximum number of non-overlapping activities using greedy approach
     * @param activities Array of activities
     * @return List of selected activities
     */
    public static List<Activity> selectActivities(Activity[] activities) {
        if (activities.length == 0) {
            return new ArrayList<>();
        }
        
        // Step 1: Sort activities by finish time (greedy choice preparation)
        Arrays.sort(activities, (a, b) -> Integer.compare(a.finishTime, b.finishTime));
        
        List<Activity> selected = new ArrayList<>();
        
        // Step 2: Always select the first activity (earliest finish time)
        selected.add(activities[0]);
        int lastFinishTime = activities[0].finishTime;
        
        // Step 3: For remaining activities, select if it starts after last selected finishes
        for (int i = 1; i < activities.length; i++) {
            // Greedy choice: if current activity starts after last selected finishes
            if (activities[i].startTime >= lastFinishTime) {
                selected.add(activities[i]);
                lastFinishTime = activities[i].finishTime;
            }
        }
        
        return selected;
    }
    
    /**
     * Alternative implementation: recursive approach
     */
    public static List<Activity> selectActivitiesRecursive(Activity[] activities) {
        if (activities.length == 0) {
            return new ArrayList<>();
        }
        
        // Sort by finish time
        Arrays.sort(activities, (a, b) -> Integer.compare(a.finishTime, b.finishTime));
        
        List<Activity> result = new ArrayList<>();
        selectActivitiesRecursiveHelper(activities, 0, -1, result);
        return result;
    }
    
    /**
     * Recursive helper for activity selection
     * @param activities Sorted array of activities
     * @param index Current index being considered
     * @param lastSelected Index of last selected activity (-1 if none)
     * @param result List to store selected activities
     */
    private static void selectActivitiesRecursiveHelper(Activity[] activities, int index, 
                                                       int lastSelected, List<Activity> result) {
        // Base case: no more activities to consider
        if (index >= activities.length) {
            return;
        }
        
        // If this activity doesn't overlap with last selected, choose it
        if (lastSelected == -1 || activities[index].startTime >= activities[lastSelected].finishTime) {
            result.add(activities[index]);
            selectActivitiesRecursiveHelper(activities, index + 1, index, result);
        } else {
            // Skip current activity and move to next
            selectActivitiesRecursiveHelper(activities, index + 1, lastSelected, result);
        }
    }
    
    /**
     * Calculate total time covered by selected activities
     */
    public static int calculateTotalTime(List<Activity> activities) {
        int totalTime = 0;
        for (Activity activity : activities) {
            totalTime += (activity.finishTime - activity.startTime);
        }
        return totalTime;
    }
    
    /**
     * Print activities in a formatted table
     */
    public static void printActivities(Activity[] activities, String title) {
        System.out.println(title);
        System.out.println("ID | Name      | Start | Finish | Duration");
        System.out.println("---|-----------|-------|--------|----------");
        
        for (Activity activity : activities) {
            int duration = activity.finishTime - activity.startTime;
            System.out.printf("%-2d | %-9s | %-5d | %-6d | %-8d\n", 
                            activity.id, activity.name, activity.startTime, 
                            activity.finishTime, duration);
        }
    }
    
    /**
     * Print selected activities with timeline
     */
    public static void printSelectedActivities(List<Activity> selected) {
        System.out.println("Selected activities (in execution order):");
        for (int i = 0; i < selected.size(); i++) {
            Activity activity = selected.get(i);
            System.out.printf("%d. %s\n", i + 1, activity);
        }
        
        System.out.println("\nTimeline visualization:");
        if (!selected.isEmpty()) {
            int maxTime = selected.get(selected.size() - 1).finishTime;
            System.out.print("Time: ");
            for (int t = 0; t <= maxTime; t += 2) {
                System.out.printf("%-3d", t);
            }
            System.out.println();
            
            for (Activity activity : selected) {
                System.out.printf("A%-2d: ", activity.id);
                for (int t = 0; t <= maxTime; t += 2) {
                    if (t >= activity.startTime && t < activity.finishTime) {
                        System.out.print("███");
                    } else {
                        System.out.print("   ");
                    }
                }
                System.out.println(" " + activity.name);
            }
        }
    }
    
    /**
     * Show why greedy choice works by comparing with other strategies
     */
    public static void demonstrateGreedyChoice(Activity[] activities) {
        System.out.println("Demonstrating why greedy choice (earliest finish time) works:");
        
        // Sort by finish time
        Activity[] byFinish = Arrays.copyOf(activities, activities.length);
        Arrays.sort(byFinish, (a, b) -> Integer.compare(a.finishTime, b.finishTime));
        
        // Sort by start time
        Activity[] byStart = Arrays.copyOf(activities, activities.length);
        Arrays.sort(byStart, (a, b) -> Integer.compare(a.startTime, b.startTime));
        
        // Sort by duration (shortest first)
        Activity[] byDuration = Arrays.copyOf(activities, activities.length);
        Arrays.sort(byDuration, (a, b) -> Integer.compare(
            a.finishTime - a.startTime, b.finishTime - b.startTime));
        
        System.out.println("1. Greedy by finish time: " + selectActivities(byFinish).size() + " activities");
        System.out.println("2. Greedy by start time: " + selectActivitiesByStart(byStart).size() + " activities");
        System.out.println("3. Greedy by duration: " + selectActivitiesByStart(byDuration).size() + " activities");
        System.out.println("Note: Finish time strategy typically gives optimal result");
    }
    
    /**
     * Helper method: select activities by start time (for comparison)
     */
    private static List<Activity> selectActivitiesByStart(Activity[] activities) {
        if (activities.length == 0) return new ArrayList<>();
        
        List<Activity> selected = new ArrayList<>();
        selected.add(activities[0]);
        int lastFinishTime = activities[0].finishTime;
        
        for (int i = 1; i < activities.length; i++) {
            if (activities[i].startTime >= lastFinishTime) {
                selected.add(activities[i]);
                lastFinishTime = activities[i].finishTime;
            }
        }
        
        return selected;
    }
    
    public static void main(String[] args) {
        System.out.println("=== Activity Selection - Greedy Algorithm ===");
        
        // Test Case 1: Classic activity selection problem
        System.out.println("Test Case 1: Classic activity selection");
        Activity[] activities1 = {
            new Activity(1, "Meeting", 1, 4),
            new Activity(2, "Workshop", 3, 5),
            new Activity(3, "Lunch", 0, 6),
            new Activity(4, "Training", 5, 7),
            new Activity(5, "Review", 3, 9),
            new Activity(6, "Planning", 5, 9),
            new Activity(7, "Demo", 6, 10),
            new Activity(8, "Testing", 8, 11),
            new Activity(9, "Deploy", 8, 12),
            new Activity(10, "Cleanup", 2, 14),
            new Activity(11, "Report", 12, 16)
        };
        
        printActivities(activities1, "Available activities:");
        System.out.println();
        
        List<Activity> selected1 = selectActivities(activities1);
        System.out.println("Maximum activities selected: " + selected1.size());
        printSelectedActivities(selected1);
        System.out.println("Total duration: " + calculateTotalTime(selected1) + " units");
        System.out.println();
        
        demonstrateGreedyChoice(activities1);
        System.out.println();
        
        // Test Case 2: All activities overlap
        System.out.println("Test Case 2: All activities overlap");
        Activity[] activities2 = {
            new Activity(1, "Task A", 0, 10),
            new Activity(2, "Task B", 2, 8),
            new Activity(3, "Task C", 4, 6),
            new Activity(4, "Task D", 1, 9)
        };
        
        printActivities(activities2, "Overlapping activities:");
        System.out.println();
        
        List<Activity> selected2 = selectActivities(activities2);
        System.out.println("Maximum activities selected: " + selected2.size());
        printSelectedActivities(selected2);
        System.out.println();
        
        // Test Case 3: No overlaps (all activities can be selected)
        System.out.println("Test Case 3: No overlapping activities");
        Activity[] activities3 = {
            new Activity(1, "Morning", 0, 2),
            new Activity(2, "Noon", 2, 4),
            new Activity(3, "Afternoon", 4, 6),
            new Activity(4, "Evening", 6, 8)
        };
        
        printActivities(activities3, "Non-overlapping activities:");
        System.out.println();
        
        List<Activity> selected3 = selectActivities(activities3);
        System.out.println("Maximum activities selected: " + selected3.size());
        printSelectedActivities(selected3);
        System.out.println();
        
        // Test Case 4: Single activity
        System.out.println("Test Case 4: Single activity");
        Activity[] activities4 = {
            new Activity(1, "Solo", 5, 10)
        };
        
        List<Activity> selected4 = selectActivities(activities4);
        System.out.println("Activities selected: " + selected4.size());
        System.out.println("Selected: " + selected4.get(0));
        System.out.println();
        
        // Test Case 5: Compare iterative vs recursive
        System.out.println("Test Case 5: Iterative vs Recursive comparison");
        List<Activity> iterativeResult = selectActivities(activities1);
        List<Activity> recursiveResult = selectActivitiesRecursive(activities1);
        
        System.out.println("Iterative approach: " + iterativeResult.size() + " activities");
        System.out.println("Recursive approach: " + recursiveResult.size() + " activities");
        System.out.println("Results match: " + (iterativeResult.size() == recursiveResult.size()));
    }
}